### PR TITLE
Replace predict feature-map dumps with Grad-CAM++ heatmaps

### PR DESCRIPTION
**Pull request by Y-T-G**
`visualize=True` in predict mode saved a PNG grid of every conv layer's feature maps plus a `.npy` per layer. It was noisy and users did not find it useful. It now saves **one Grad-CAM++ heatmap per image**, showing which pixels drove the predicted class scores.

`visualize=True` in **val** mode is untouched. It still draws TP, FP and FN.

**Three changes:**

1. **`feature_visualization` is replaced by `gradcam`** in `ultralytics/utils/plotting.py`. It hooks the feature maps entering the model head, backprops the class logits of the predictions above `conf`, and saves `<image>_cam.jpg` next to the run. It returns the same predictions, so there is no second forward pass.
2. **The per-layer `visualize` hook is deleted from the four model forward loops** in `nn/tasks.py`, together with the now dead `visualize` argument on `BaseModel.predict`, `BaseModel._predict_once`, `Ensemble.forward`, `AutoBackend.forward` and `PyTorchBackend.forward`.
3. **`Detect` only detaches the one2one branch during training.** At inference the detach did nothing, and it blocked gradients from reaching the neck features on end-to-end models like YOLO26.

**Why Grad-CAM++ and not Grad-CAM or Eigen-CAM:** Grad-CAM++ weights each spatial position on its own, so images with many objects of the same class keep every instance visible. I compared Grad-CAM, HiResCAM, LayerCAM and Eigen-CAM on the same images. The others collapsed onto a single object or lit up most of the background. Two details matter for YOLO: the target uses the raw class logits, not the sigmoid scores, and activations are clamped at 0 in the position weight term because the derivation assumes non-negative activations while SiLU can go negative. The maps of all head levels (P3, P4, P5) are summed, so small and large objects share one overlay.

**Tasks tested:** detect, segment, pose, obb, classify, semantic, depth, grayscale, YOLO-World, YOLOE and RT-DETR, plus batch sizes above 1.

**Not supported, falls back to normal inference:** non-PyTorch formats, tensor sources, `augment=True` and `embed`. None of these produced feature maps before either.

| | before | after |
|---|---|---|
| files per image | 1 PNG + 1 NPY per conv layer | 1 JPG |
| shows | raw channel activations | pixels behind the predictions |
| needs matplotlib | yes | no |

Deleted: `ultralytics.utils.plotting.feature_visualization`

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Replaces predict-time feature-map dumps with Grad-CAM++ heatmaps that highlight image regions influencing model predictions. ✨

### 📊 Key Changes
- Adds a PyTorch Grad-CAM++ utility that captures head activations and class-score gradients, combines feature-pyramid levels, and saves heatmap overlays beside predictions.
- Updates `visualize` documentation and defaults to describe Grad-CAM++ output, including PyTorch-only support.
- Removes the legacy feature-map visualization path and its `visualize` arguments from model prediction and backend forwarding APIs.
- Disables Grad-CAM++ for tensor sources, augmented inference, embeddings, and non-PyTorch backends.
- Preserves gradient flow through end-to-end detection heads during evaluation so explainability maps can be generated.
- Adds a COCO8 NDJSON example dataset file for visualization or related workflows. 📦

### 🎯 Purpose & Impact
- Produces more actionable, prediction-focused visual explanations than raw intermediate feature-map dumps.
- Saves one Grad-CAM++ overlay per input image, helping users understand which pixels contributed to detected classes.
- Maintains standard inference behavior for unsupported configurations and non-PyTorch models.
- Updates user-facing documentation and API behavior; consumers relying on the removed internal `visualize` feature-map interface may need to migrate.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your Grad-CAM++ visualization work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for the new `visualize` behavior and Grad-CAM++ support.
- ✅ **Add Tests**: If applicable, include or update tests to cover heatmap generation, supported and unsupported inference paths, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). This is an automated welcome message, and an Ultralytics engineer will assist with review or any questions. Thank you for contributing to Ultralytics! 🚀

**Comment by Y-T-G** [comment]
Reopened as #25548 from a branch on ultralytics/ultralytics instead of the fork.

File: ultralytics/utils/plotting.py [hunk 1]
```diff
@@ -1304,37 +1304,86 @@ def _save_one_file(file):
...
+        raw = output[1] if isinstance(output, tuple) else output  # heads returning (predictions, raw) keep the raw
+        if isinstance(raw, dict):  # Detect and subclasses, end2end heads predict from their one2one branch
+            s = raw.get("one2one", raw)["scores"]  # (B, nc, anchors)
+        elif isinstance(raw, tuple):  # RTDETRDecoder, raw = (dec_bboxes, dec_scores, ...)
+            s = raw[1].squeeze(0).transpose(1, 2)  # (B, nc, queries)
+        else:  # Classify (B, nc), SemanticSegment (B, nc, h, w), Depth (B, 1, h, w)
+            s = raw
+        scores.append(s.reshape(*s.shape[:2], -1))  # class logits, (B, nc, predictions)
+
+    head = model.model.model[-1]  # AutoBackend -> PyTorch model -> head
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/plotting.py:1357]: 💡 **MEDIUM**: This assumes every native PyTorch model exposes a final head through `.model[-1]`. A supported `Ensemble` does not: `AutoBackend.model` resolves to the `Ensemble`, which is a `ModuleList` with no `.model` attribute, so `predict(..., visualize=True)` raises `AttributeError` before inference. Handle ensembles explicitly (for example, choose/register each member head) or reject visualization for them before calling `gradcam`.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The Grad-CAM++ path is well-contained and preserves the normal prediction path, but it currently regresses `visualize=True` for native PyTorch ensembles due to an invalid head lookup.

💬 Posted 1 inline comment
